lib/Backends/OpenCL/OpenCL.cpp

```diff
@@ -1452,141 +1441,181 @@ uint64_t OpenCLFunction::copyValueFromDevice(const Value *v, void *buf) {
+      kernelLaunches_.emplace_back(KernelLaunch("copyToDevice", event));
     }
-    copiedBytes += copyValueToDevice(it.first);
   }
   // Do it!
   clFinish(commands_);
-  return copiedBytes;
 }
 
-uint64_t OpenCLFunction::copyMutableWeightsFromDevice() {
-  size_t copiedBytes = 0;
-  clFinish(commands_);
-
-  for (auto it : tensors_) {
-    if (!externalTensors_.count(it.first)) {
-      continue;
-    }
-    if (auto *W = dyn_cast<WeightVar>(it.first)) {
-      if (W->getMutability() == WeightVar::MutabilityKind::Constant)
-        continue;
+void OpenCLFunction::copyOutputsFromDevice(const Context &ctx) {
+  for (auto PH : ctx.pairs()) {
+    auto symbolInfo =
+        bundle_.symbolTable.find(std::string(PH.first->getName()));
+    assert(symbolInfo != bundle_.symbolTable.end() && "Symbol not found");
+    auto addr = symbolInfo->second.offset;
+    auto numBytes = symbolInfo->second.size;
+    // Issue a non-blocking command to copy the buffer to the device.
+    auto buf = PH.second->getUnsafePtr();
+    cl_event event{nullptr};
+    cl_int err = clEnqueueReadBuffer(
+        commands_, deviceBuffer_, /* blocking_read */ CL_FALSE, addr, numBytes,
+        buf, /* num_events_in_wait_list */ 0,
+        /* event_list */ nullptr, /* event */ doProfile ? &event : nullptr);
+    GLOW_ASSERT(err == CL_SUCCESS && "Unable to copy data to the device");
+    if (doProfile) {
+      kernelLaunches_.emplace_back(KernelLaunch("copyToDevice", event));
     }
-    copiedBytes += copyValueFromDevice(it.first);
   }
+  // Do it!
   clFinish(commands_);
-  return copiedBytes;
 }
-
-void OpenCLFunction::allocateMemory(const Context &ctx) {
-  // The allocator assigns device memory addresses to the buffers.
+/// Collect Contstants from graph and store them for passing to execution.
+uint8_t *collectConstants(
+    const IRFunction *F, uint64_t constantMaxSize,
+    std::unordered_map<std::string, runtime::RuntimeSymbolInfo> symbolTable) {
+
+  // At compile time condense constants to a single block of memory.
+  // This allows the graph to go away after compile time.
+  uint8_t *baseConstantWeightVarsStore =
+      (uint8_t *)alignedAlloc(constantMaxSize, TensorAlignment);
+  for (auto &v : F->getGraph()->getParent()->getConstants()) {
+    assert(isa<WeightVar>(F->getWeightForNode(v)));
+    auto *w = cast<WeightVar>(F->getWeightForNode(v));
+    auto payload = v->getPayload().getUnsafePtr();
+    auto numBytes = w->getSizeInBytes();
+    auto addr = symbolTable[std::string(w->getName())].offset;
+    // Copy weight to offset.
+    memcpy(baseConstantWeightVarsStore + addr, payload, numBytes);
+  }
+  return baseConstantWeightVarsStore;
+}
+/// Computes offsets and total allocation for Constants, Placeholders, and
+/// Activations to build runtime symbol table. Returns
+/// RuntimeBundle.
+runtime::RuntimeBundle generateRuntimeBundle(const IRFunction *F) {
+  // Use a single allocator. The OpenCL backend uses a single buffer on the card
```

Comment:
IIRC, you also have introduced a similar `generateRuntimeBundle` in the CPU backend. May be you should create a utility function and call it from both backends?

Reply:
That function differs a fair bit between backends. For now I recommend keeping them separate.